[#77] 공연(좌석) 예매 기능 구현

hanwix2 opened the pull request:
#77 

## 개요
- **서비스 시나리오:**
    1. 예매하고자 하는 공연에 대한 특정 스케줄의 전체 좌석 정보를 먼저 조회.
    2. 조회된 좌석 정보에는 좌석 위치와 가격, 예매 여부가 포함
    3. 원하는 좌석을 선택(다중 선택 가능)하여 예매 요청
- **특이사항:**
    - 한번에 최대 예매 가능 좌석수는 4자리
    - 다른 스케줄 또는 공연의 좌석 동시 예매 불가능
    - *결제 시스템 연동*은 추후에 적용 예정
- *일반 회원 기능*


## Progress

- [x] 기능 구현
- [x] 테스트 코드

---
### 기능 구현
- **DB에 예매(reservation) 테이블 생성**
    - 속성: id, userId, seatId, reserveDate(예매날짜)

<br>

- **공연 예매 기능**
    1. ***ReservationRequest* 객체를 전달받아 예매 요청**
        - 인스턴스: perfTimeId(공연 스케줄 id), seatIds(예매할 좌석id 리스트)
    2. **공연 예매 처리 전 공연 시간이 존재하는지 확인**
    3. **예매 가능한 좌석 정보인지 확인**(checkSeatsReservable())
        - 체크 항목: 
            - 유효한 좌석id
            - 예매 가능 여부
            - 동일한 스케줄
        - 요청을 보낸 좌석수와 DB에서 예매 가능한 좌석수를 비교하여 판단
    4. **좌석(seat) 데이터에 예매상태(reserved) true로 변경** (setSeatsReserved())
    5. **예매(reservation) 테이블에 예매 데이터 추가** (reserveSeats())

<br>

- 트랜잭션 적용

<br> 

### 예매가 되지 않는 상황
- **최대 예매 가능 좌석수(4)를 초과**하여 요청한 경우
- 요청 데이터의 **공연 스케줄이 존재하지 않는** 경우
- **존재하지 않는 좌석**이 포함된 경우
- **이미 예매된 좌석**이 포함된 경우
- 요청한 좌석이 모두 **동일한 스케줄이 아닌 경우**

hanwix2 committed:
[#77] 공연 예매 기능 구현

- DB 예매(reservation) 테이블 생성
    - 속성: id, userId, seatId, reserveDate(예매날짜)
- 공연 예매(reservation) 관려 컨트롤러, 서비스, 매퍼 클래스 생성
- 공연 예매 기능
    - ReservationRequest 객체를 전달받아 예매 요청
        - 인스턴스: perfTimeId(공연 스케줄 id), seatIds(예매할 좌석id 리스트)
        - 한번에 여러 좌석 예매 가능
    - 공연 예매 처리 전 공연 시간이 존재하는지 확인
    - setSeatsReserved(): 좌석(set) 데이터에 예매상태(reserved) true로 변경
    - reserveSeats(): 예매(reservation) 테이블에 예매 데이터 추가
    - 트랜잭션 적용

hanwix2 committed:
[#77] 최대 예매 가능 좌석 수 지정

- 소수의 회원이 많은 좌석을 독점하는 것을 방지하기 위해 단일 요청당 최대 예매 가능 좌석 수를 4개로 지정
- 만약 허용 예매 좌석 수를 넘겨 예매 요청시 ReserveAllowedQuantityExceededException 예외 처리

hanwix2 committed:
[#77] 좌석 예매가능 여부 확인 로직 추가

- DB에서 예매가능한 좌석 수와 요청한 좌석 수를 비교하여 예매 가능 여부 판단
- 예매 불가능한 경우:
    1. 존재하지 않는 좌석이 포함된 경우
    2. 이미 예매된 좌석이 포함된 경우
    3. 요청한 좌석이 모두 동일한 스케줄이 아닌 경우

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/ReservationService.java:
```diff
@@ -0,0 +1,42 @@
+@RequiredArgsConstructor
+public class ReservationService {
+
+    private final static int MAX_RESERVATION_QUANTITY = 4;
```
이런 비즈니적인 설정값은 소스코드에 넣어두기보다는 `@Value`를 통해 외부에서 주입받도록 하면 좋을 것 같아요~

hanwix2 committed:
[#77] 예매 기능에 대한 테스트 코드 추가

hanwix2 committed:
[#77] MAX_RESERVATION_QUANTITY 값 외부에서 주입

- MAX_RESERVATION_QUANTITY: 최대 예매 가능 좌석 수 (int형 상수)
- properties 파일에서 값 지정 및 @Value를 통한 주입
- @Value 어노테이션을 이용한 상수화를 위해 setter 적용

f-lab-dev commented on src/test/java/com/show/showticketingservice/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+        }
+
+        @Test
+        @DisplayName("[예매 좌석 수: 5 / 실패] 최대 예매 가능 좌석 수(4개)를 초과하여 예매할 경우")
```
`[예매 좌석 수: 5 / 실패]` 처음보는 사람의 입장에서는 이 내용이 무슨 말인지 알아들을 수 있을까요? 사실 테스트 코드 내용을 보면 알 수 있긴 하지만 바로 알아볼 수 있도록 글을 써보는것도 중요합니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/ReservationService.java:
```diff
@@ -14,14 +15,19 @@
 public class ReservationService {
 
-    private final static int MAX_RESERVATION_QUANTITY = 4;
+    private static int MAX_RESERVATION_QUANTITY;
```
static을 굳이 써야할까요?

hanwix2 replied:
지금까지 상수화는 일반적으로 `static final`로 선언한다는 것에 너무 집중한 것 같습니다.

상수화를 할 때 `static final`을 사용하는 이유는 해당 클래스를 사용하는 모든 곳에서 동일한 값을 유지하기 위해서인데, 

spring에서는 빈으로 관리가 되어 **하나의 객체만 만들어지기 때문에** 굳이 static을 사용할 이유가 없다는 것을 간과했습니다. 

그러므로 `final`로만 선언하여 상수로써 immutable 특성만 가지면 될 것 같습니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/ReservationService.java:
```diff
@@ -14,14 +15,19 @@
     private final SeatService seatService;
 
+    @Value("${service.value.max-reservation-quantity}")
+    public void setMaxReservationQuantity(int quantity) {
```
setter 말고 스프링에서 권장하는 주입방식이 있었던 것 같습니다~

hanwix2 replied:
위의 답변에 이어지는 내용입니다.

static final 로 선언한 상수를 외부에서 주입하는 것은 불가능했습니다. 그래서 static 변수를 초기화 하는 것에 집중하다보니 setter를 사용했던 것 같네요. 

static으로 선언할 필요가 없으므로

**생성자 주입**을 이용해서 private final 변수를 초기화 하면 될 듯합니다.

hanwix2 committed:
[#77] ReservationService의 상수 생성자 주입 적용

- 상수: MAX_RESERVATION_QUANTITY(최대 예매 가능 좌석 수)
- static이 아닌 private final 으로 선언
- 테스트 코드 반영

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/ReservationService.java:
```diff
@@ -4,28 +4,31 @@
 public class ReservationService {
 
-    private static int MAX_RESERVATION_QUANTITY;
+    private final int MAX_RESERVATION_QUANTITY;
```
static이 아니므로 카멜케이스를 지켜주세요~

f-lab-dev reviewed (commented):
static 변수를 setter를 이용해 수정하는 행위도 위험합니다~ 왜 그런지 고민해보시면 좋겠습니다

hanwix2 committed:
[#77] 예매 관련 상수명 수정

static을 사용하지 않은 상수명에 camel case 적용

hanwix2 committed:
[#116] Jenkinsfile 생성

hanwix2 committed:
[#116] unit test를 위한 surefire plugin 설정

- surefire를 이용한 단위 테스트 수행 시 외부 환경에 종속적인 테스트가 포함된 Controller 테스트 제외

hanwix2 committed:
[#116] 단위 테스트를 위한 pom.xml 수정

hanwix2 committed:
Merge pull request #117 from f-lab-edu/feature/116/jh

[#116] Jenkins CI 적용
